### PR TITLE
CHE-7156: Replace SearchEngine in Java debugger

### PR DESCRIPTION
**Pull request by mmorhun**
### What does this PR do?
Replaces `SearchEngine` which was used for finding resource by an FQN.
This change is made because of issues with debugger performance.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/7156

### CODE REVIEW

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelUtil.java [hunk 1]
```diff
@@ -59,6 +53,19 @@ public static IJavaProject getJavaProject(String resourceUri) {
...
 
+  /**
+   * Returns all user created java project which exist in current workspace. This method excludes
+   * default {@code jdt.ls-java-project} project.
+   *
+   * @return all user created java projects in current workspace
+   */
+  public static List<IJavaProject> getWorkspaceJavaProjects() {
+    return Arrays.stream(ResourcesPlugin.getWorkspace().getRoot().getProjects())
+        .filter(project -> !project.getName().equals("jdt.ls-java-project"))
```
Comment (svor) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelUtil.java]: As for me would be better if `jdt.ls-java-project`  would be declared as constant. And for equals operation use a constant firstly than an object with which to compare to avoid NPE.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelUtil.java [hunk 2]
```diff
@@ -22,18 +24,12 @@
...
 public class JavaModelUtil {
-  private static final Gson gson =
-      new GsonBuilder()
-          .registerTypeAdapterFactory(new CollectionTypeAdapterFactory())
-          .registerTypeAdapterFactory(new EitherTypeAdapterFactory())
-          .registerTypeAdapterFactory(new EnumTypeAdapterFactory())
-          .create();
+  public static final String JDT_LS_JAVA_PROJECT = "jdt.ls-java-project";
+
+  private static final Gson gson = GsonUtils.getInstance();
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/JavaModelUtil.java]: +1

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java [hunk 3]
```diff
@@ -178,30 +174,32 @@ public static String identifyFqnInResource(List<Object> params, IProgressMonitor
...
    */
   public static List<Either<String, ResourceLocation>> findResourcesByFqn(
-      List<Object> params, IProgressMonitor pm) {
+      List<Object> params, IProgressMonitor progressMonitor) {
     Preconditions.checkArgument(params.size() >= 1, "FQN is expected.");
 
+    ensureNotCancelled(progressMonitor);
+
     String fqn = (String) params.get(0);
+    List<IJavaProject> projects = getWorkspaceJavaProjects();
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java]: Unfortunately, we don't control the order or lookup here, but this is a deeper problem that is not part of this PR.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java [hunk 4]
```diff
@@ -178,30 +174,32 @@ public static String identifyFqnInResource(List<Object> params, IProgressMonitor
...
+      }
     }
 
-    if (types.isEmpty()) {
-      throw new RuntimeException("Type with fully qualified name: " + fqn + " was not found");
-    }
+    ensureNotCancelled(progressMonitor);
 
-    List<Either<String, ResourceLocation>> result = new ArrayList<>();
+    List<Either<String, ResourceLocation>> result = new ArrayList<>(types.size());
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/debug/FqnDiscover.java]: Can't we simply return "JDTUtils.getURI(type.getTypeRoot()) ?
